src/Categories/Category/Monoidal/Interchange.agda

```diff
@@ -0,0 +1,603 @@
+                         σ⇒ ⊗₁ σ⇒   ⇒⟨ (Y ⊗₀ W) ⊗₀ (Z ⊗₀ X) ⟩
+                         i⇒
+                       ≈ σ⇒ 
+                       ⟩
+  swapInner-braiding = begin
+      i⇒ ∘ σ⇒ ⊗₁ σ⇒ ∘ i⇒
+    ≡⟨⟩
+      i⇒ ∘ σ⇒ ⊗₁ σ⇒ ∘ α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒
+    ≈⟨ swapInner-coherent ⟩∘⟨ pushˡ serialize₁₂ ⟩
+      i′⇒ ∘ σ⇒ ⊗₁ id ∘ id ⊗₁ σ⇒ ∘ α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒
+    ≈˘⟨ ((refl⟩∘⟨ refl⟩⊗⟨ ⊗.identity) ⟩∘⟨ ⊗.identity ⟩⊗⟨refl ⟩∘⟨refl) ○ assoc ⟩
+      ((α⇒ ∘ (α⇐ ∘ id ⊗₁ σ⇒ ∘ α⇒) ⊗₁ id ∘ α⇐) ∘ σ⇒ ⊗₁ (id ⊗₁ id)) ∘
+      (id ⊗₁ id) ⊗₁ σ⇒ ∘ α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒
+    ≈⟨ pullʳ (pullʳ assoc-commute-to) ⟩∘⟨ extendʳ (⟺ assoc-commute-to) ⟩
+      (α⇒ ∘ (α⇐ ∘ id ⊗₁ σ⇒ ∘ α⇒) ⊗₁ id ∘ (σ⇒ ⊗₁ id) ⊗₁ id ∘ α⇐) ∘
+      α⇐ ∘ id ⊗₁ (id ⊗₁ σ⇒) ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒
+    ≈˘⟨ (refl⟩∘⟨ pushˡ split₁ˡ) ⟩∘⟨ refl⟩∘⟨ pushˡ split₂ˡ ⟩
+      (α⇒ ∘ ((α⇐ ∘ id ⊗₁ σ⇒ ∘ α⇒) ∘ σ⇒ ⊗₁ id) ⊗₁ id ∘ α⇐) ∘
+      α⇐ ∘ id ⊗₁ (id ⊗₁ σ⇒ ∘ α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒
+    ≈⟨ (refl⟩∘⟨ pullʳ (assoc ○ hexagon₁) ⟩⊗⟨refl ⟩∘⟨refl) ⟩∘⟨
+        refl⟩∘⟨ refl⟩⊗⟨ (⟺ assoc ○ (pullˡ (assoc ○ hexagon₁))) ⟩∘⟨refl ⟩
+      (α⇒ ∘ (α⇐ ∘ α⇒ ∘ σ⇒ ∘ α⇒) ⊗₁ id ∘ α⇐) ∘
+      α⇐ ∘ id ⊗₁ ((α⇒ ∘ σ⇒ ∘ α⇒) ∘ α⇐) ∘ α⇒
+    ≈⟨ (refl⟩∘⟨ cancelˡ α.isoˡ ⟩⊗⟨refl ⟩∘⟨refl) ⟩∘⟨
+        refl⟩∘⟨ refl⟩⊗⟨ pullʳ (cancelʳ α.isoʳ) ⟩∘⟨refl ⟩
+      (α⇒ ∘ (σ⇒ ∘ α⇒) ⊗₁ id ∘ α⇐) ∘ α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒) ∘ α⇒
+    ≈⟨ extendʳ (pushʳ split₁ˡ) ⟩∘⟨ extendʳ (pushʳ split₂ˡ) ⟩
+      ((α⇒ ∘ σ⇒ ⊗₁ id) ∘ α⇒ ⊗₁ id ∘ α⇐) ∘ (α⇐ ∘ id ⊗₁ α⇒) ∘ id ⊗₁ σ⇒ ∘ α⇒
+    ≈˘⟨ extendʳ (pushʳ (pushʳ assoc)) ⟩
+      (α⇒ ∘ σ⇒ ⊗₁ id) ∘ (α⇒ ⊗₁ id ∘ (α⇐ ∘ α⇐) ∘ id ⊗₁ α⇒) ∘ id ⊗₁ σ⇒ ∘ α⇒
+    ≈˘⟨ refl⟩∘⟨ switch-tofromˡ (α ⊗ᵢ idᵢ)
+                  (switch-tofromʳ (idᵢ ⊗ᵢ α) inv-pentagon) ⟩∘⟨refl ⟩
+      (α⇒ ∘ σ⇒ ⊗₁ id) ∘ α⇐ ∘ id ⊗₁ σ⇒ ∘ α⇒
+    ≈⟨ pullʳ (sym-assoc ○ pullˡ hexagon₂) ⟩
+      α⇒ ∘ ((α⇐ ∘ σ⇒) ∘ α⇐) ∘ α⇒
+    ≈⟨ (refl⟩∘⟨ cancelʳ α.isoˡ) ⟩
+      α⇒ ∘ (α⇐ ∘ σ⇒)
+    ≈⟨ cancelˡ α.isoʳ ⟩
+      σ⇒
+    ∎
+    where inv-pentagon = _≃_.to-≈ pentagon-iso
+
+  -- Braided monoidal categories have an interchange.
+
+  hasInterchange : HasInterchange M
+  hasInterchange = record
+    { swapInner         = swapInner
+    ; swapInner-natural = swapInner-natural
+    ; swapInner-assoc   = swapInner-assoc
+    ; swapInner-unitˡ   = swapInner-unitˡ
+    ; swapInner-unitʳ   = swapInner-unitʳ
+    }
+  open HasInterchange hasInterchange public using (swapInner-naturalIso)
+
+-- Extra identities that hold only for symmetric monoidal categories.
+
+module SymmetricInterchange {M : Monoidal C} (S : Symmetric M) where
+  open MonoidalReasoning M
+  open MonoidalUtilities M
+  open Symmetric S renaming (associator to α)
+  open Shorthands
+  open BraidedInterchange braided public
+  private
+    i⇒ = swapInner.from
+    i⇐ = swapInner.to
+    σ⇒ = λ {X Y} → braiding.⇒.η (X , Y)
+    σ⇐ = λ {X Y} → braiding.⇐.η (X , Y)
+
+  swapInner-commutative : ∀ {W X Y Z} → i⇒ ∘ i⇒ {W} {X} {Y} {Z} ≈ id
```

Comment:
I think it will be clearer if we make objects generalized variables.

Reply:
OK, I can try that out. But I think the explicit arguments will still be needed here. We'll just get rid of the quantified ones. Maybe a better way would be to use `Commutation`? WDYT?

Reply:
I agree that in this exact spot, generalized variables might not work so well, because we want to know which order they are specified in (i.e. we might need to give them to Agda, and if we don't know the order, code breaks).

I forget about `Commutation`, can you remind me?

Reply:
I mean the syntax used to specify the compositions as 'sequences' of arrows with explicit domains and codomains. I use these in the signature of e.g. `swapInner-assoc`.

Reply:
Ah right. It would make the whole type clearer, which would be a nice piece of documentation.

Reply:
OK, I'll fix that.

Reply:
Question: I see that all the `variable` declarations in the library are in `private` blocks. Why is that? How does `private` modify `variable`?

Reply:
`variable` can be exported. so we use `private` to prevent this from leaking out of a module. a convention in stdlib.

Reply:
OK, makes sense. Thanks!

I updated the file now so it uses variables. And also added the richer type signature for this particular case. Hope that's better.